File: src/main/scala/org/bakkes/game/view/components/LineWriter.java
package org.bakkes.game.view.components;

import java.util.LinkedList;
import java.util.List;

import org.bakkes.game.view.IRenderable;
import org.newdawn.slick.Color;
import org.newdawn.slick.Font;
import org.newdawn.slick.Graphics;

import com.google.inject.Inject;
import com.google.inject.Provider;

/**
 * don't extends a view otherwise stackoverflow
 *
 * a simple utility to emulate the cout << "" kind of stuff on screen
 */
public class LineWriter extends APositionable implements IRenderable{

	private List<TextLine> lines = new LinkedList<>();
	private @Inject Provider<TextLine> textLineProvider;
	public Color color = Color.black;

	@Override
	public void render(final Graphics g) {
		g.setColor(color);
		for(final TextLine l : lines){
			l.render(g);
		}
	}

	public float getHeight(){
		float lowY = Float.POSITIVE_INFINITY;
		float highY = Float.NEGATIVE_INFINITY;
		for(final TextLine t : lines){
			if(lowY > t.y()){
				lowY = t.y();
			}
			if(highY < t.y()){
				highY = t.y();
			}
		}
		if(lowY == Float.POSITIVE_INFINITY || highY == Float.NEGATIVE_INFINITY){
			return 0;
		}
		return highY - lowY;
	}

	public void write(final String str){
		write(str, null);
	}
	public void write(final String str, final Font font){
		final TextLine t = textLineProvider.get();
		t.setText(str);
		t.y(y());
		t.x(x());
		t.setFont(font);
		write(t);
	}
	public void write(final TextLine line){
		lines.add(line);
		y(y()+line.height());
	}
	public void clear(){
		lines.clear();
	}
	public int lineCount(){
		return lines.size();
	}
}

File: src/main/scala/org/bakkes/game/view/components/LineWriter.scala
package org.bakkes.game.view.components

import com.google.inject.{Inject, Provider}
import org.bakkes.game.view.IRenderable
import org.newdawn.slick.{Color, Font, Graphics}

import scala.collection.immutable.List

/**
 * works like a type writer for shape, it will go down everytime a shape is added
 * by the height of it, originaly intended only for strings, but after observation
 * it works for anything with the dimensions of a shape
 *
 * don't extends a view otherwise stackoverflow
 *
 * a simple utility to emulate the cout << "" kind of stuff on screen
 */
class LineWriter @Inject() (
		private var textLineProvider: Provider[ITextableShape]
						 )extends AShape with IRenderable {
	private var lines = List[IShape]()
	var color: Color = Color.black

	def render(g: Graphics) {
		g.setColor(color)
		for (l <- lines) {
			l.render(g)
		}
	}

	def width: Float = {
		var result = 0f
		val max = lines.sortWith(
			(a, b) => a.x() + a.width() < b.x() + b.width()
		).head
		val min = lines.sortBy((x) => x.x()).last
		result = max.x - min.x
		return result
	}

	def height: Float = {
		val sorted = lines.sortBy((y) => y.y())
		return sorted.last.y() - sorted.head.y()
	}

	def write(str: String) {
		write(str, null)
	}

	def write(str: String, font: Font) {
		val t = textLineProvider.get
		t.setText(str)
		t.y(y)
		t.x(x)
		t.setFont(font)
		write(t)
	}

	def write(line: IShape) {
		lines =  line :: lines
		y(y + line.height)
	}

	def clear = lines = List[IShape]()
	def lineCount: Int = lines.count((x)=>true)
}